core/runtime/src/main/java/io/quarkus/runtime/CracRecorder.java

```diff
@@ -0,0 +1,95 @@
+package io.quarkus.runtime;
+
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.net.URLConnection;
+import java.util.Enumeration;
+
+import org.crac.Context;
+import org.crac.Resource;
+import org.jboss.logging.Logger;
+
+import io.quarkus.runtime.annotations.Recorder;
+
+/**
+ * Registers a CRAC resource. Must be called in static initialization phase!
+ */
+@Recorder
+public class CracRecorder {
+
+    protected static class CracResource implements Resource {
+        protected final boolean fullWarmup;
+        protected final boolean preloadClasses;
+        protected int numPreloaded;
+
+        public CracResource(boolean preloadClasses, boolean fullWarmup) {
+            this.preloadClasses = preloadClasses;
+            this.fullWarmup = fullWarmup;
+        }
+
+        @Override
+        public void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+            if (preloadClasses) {
+                preloadClasses();
+            }
+            if (fullWarmup) {
```

Comment:
I don't feel great about this, as anything with sockets will certainly break.

Reply:
It's either have a warmup phase or do start() within INIT phase of CRAC and reconnect on the fly.  That's what Amazon is suggesting and what they are saying everybody else does.